Add a Dev UI Screen for Agroal (DB)

phillip-kruger opened the pull request:
This PR adds a Dev UI screen for Agroal (so any time you have a DB). 

![datasource](https://github.com/user-attachments/assets/c2f928c9-261d-40eb-9a26-9177c69aa0b9)

You can view a more detailed demo here: https://youtu.be/-ms_2ayumkk

gsmet commented on extensions/agroal/deployment/src/main/resources/dev-ui/qwc-agroal-datasource.js:
```diff
@@ -0,0 +1,339 @@
+    }
+    
+    _clearSqlInput(){
+        this._executeSQL("select * from " + this._selectedTable.tableName);
+        this._fetchAllData();
```
How will it scale when you have a lot of data?
That could happen even in Dev UI as you might have an already populated database. At least, that's how we worked at my previous company once the initial development was over, we were working daily with a fully populated dataset coming from the production.

phillip-kruger replied:
I limit on the server to max 100. The reason I am not doing it here (in Sql) is the SQL is different per vendor. So I just do it in the result set

mkouba replied:
Then the UI should reflect this somehow, or?

yrodiere replied:
Related: https://github.com/quarkusio/quarkus/pull/43618#discussion_r1782441706

gsmet commented on extensions/agroal/deployment/src/main/resources/dev-ui/qwc-agroal-datasource.js:
```diff
@@ -0,0 +1,339 @@
+    _renderTableData(){
+        if(this._selectedTable && this._currentDataSet && this._currentDataSet.cols){
+            return html`<div class="data">
+                            <vaadin-grid .items="${this._currentDataSet.data}" theme="row-stripes no-border" class="fill" column-reordering-allowed>
```
I'm wondering what happens if you have a `bytea` in your table? Or some text that is very large.

FroMage replied:
I have a lot of `byte[]` in my entities :) Images, mostly.

gsmet reviewed (commented):
The idea is nice and I can definitely see the value but I wonder if it won't open a can of worms. Data is hard and displaying the data coming directly from a table is not that easy - see my comments inline for a few cases where things might go wrong.

Now, what I don't know is if it's just a matter of tweaking things a bit or if it will be a continuous stream of corner cases to fix.

/cc @yrodiere

yrodiere commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/devui/DatabaseInspector.java:
```diff
@@ -0,0 +1,174 @@
+
+                        // Get the data
+                        List<Map<String, String>> rows = new ArrayList<>();
+                        while (resultSet.next() && rows.size() < 100) { // Linit to 100
```
Note some JDBC drivers may fetch the whole result set in memory even if you only access the first few rows... this could lead to `OutOfMemoryError` if we're not careful.

Things like `java.sql.Statement#setFetchSize` can help -- though even that is just a hint, and IIRC the MySQL JDBC driveer will ignore it unless it's set to `Integer.MIN_VALUE`...

maxandersen replied:
yes, result set only filtering is not going to work for.

is scrollableResults honored by most drivers these days? maybe that could work?

yrodiere replied:
> is scrollableResults honored by most drivers these days? maybe that could work?

`ScrollableResults` is a Hibernate ORM feature. Not available using just JDBC.

As far as I know, with JDBC you can only use `ResultSet`, and like I said behavior varies depending on the JDBC driver.

maxandersen reviewed (dismissed):
first all - love the feature and something I did some early experiments on (but without the ui skills :)

I have concerns though that is similar to @gsmet and something we at least should consider before pushing in, in no particular order:

1) which schema are you listing? the default the connection is for or all of them ?
with JPA on top users can very specifically limit what data is ever exposed but here its all raw...

2) data security - we already have concerns about exposing config data in devui; but now we will potentially expose a query interface to *all* data in the raw jdbc connection on the agroal connection? ...should we turn this off by default or at least have option to turn on? (maybe separate devmode only extension?)

3) feature creep - this will open up for massive set of feature requests and can keep us busy forever. We should either agree to deliberately keep this very simple or explore if more suitable reuse of some existing webbased db ui ?

yrodiere reviewed (commented):
Thanks, looking good -- though I added a few comments below.

> This PR adds a Dev UI screen for Agroal (so any time you have a DB).

Note Agroal is not necessarily present if you have a reactive datasource. You might need a similar Dev UI for reactive SQL clients...

> Now, what I don't know is if it's just a matter of tweaking things a bit or if it will be a continuous stream of corner cases to fix.

I suspect there will be lots of corner cases, though whether we get to hear about it will depend on the success of this feature.

I know that _I_ would rather work on #39584 than fix an endless stream of bugs related to JDBC quirks -- Hibernate at least hides some of them.

But I agree this is a (very!) nice start, so perhaps we could merge this as a "preview" feature, and depending on feedback we'll decide whether we keep it as is, or try to migrate it to something more "Hibernate-focused" as part of #39584? E.g. if data types are really too hard to handle, Hibernate could help (a lot).

FroMage commented:
Looks like a great feature! Now, as @yrodiere said, I want to be able to run HQL queries, not just SQL ones ;)
Oh, and also click from the table definition to the entity in the IDE?
But congrats, this is looking great!

maxandersen commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/devui/DatabaseInspector.java:
```diff
@@ -0,0 +1,174 @@
+        return tableList;
+    }
+
+    public DataSet executeSQL(String datasource, String sql) {
```
do i read it right that the client side can request arbitrary SQL execution here and not just limited to browse the known tables? is that not too broad security wise ?

phillip-kruger replied:
Yes you read it right :)

maxandersen replied:
Ok - so shouldn't we restrict this somehow ? Do we really need to expose arbitrary execution all the way out to the client ?

phillip-kruger replied:
I'll replay soon with some updates. We will hide this behind config

phillip-kruger commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/devui/DatabaseInspector.java:
```diff
@@ -0,0 +1,174 @@
+        List<Table> tableList = new ArrayList<>();
+        try {
+            AgroalDataSource ads = (AgroalDataSource) checkedDataSources.get(datasource);
+            try (Connection connection = ads.getConnection()) {
+                DatabaseMetaData metaData = connection.getMetaData();
+
+                // Get all tables
+                try (ResultSet tables = metaData.getTables(null, null, "%", new String[] { "TABLE" })) {
+                    while (tables.next()) {
+                        String tableName = tables.getString("TABLE_NAME");
+                        String tableSchema = tables.getString("TABLE_SCHEM");
+
+                        // Get the Primary Keys
+                        List<String> primaryKeyList = new ArrayList<>();
+                        try (ResultSet primaryKeys = metaData.getPrimaryKeys(null, tableSchema, tableName)) {
+                            while (primaryKeys.next()) {
+                                String primaryKeyColumn = primaryKeys.getString("COLUMN_NAME");
+                                primaryKeyList.add(primaryKeyColumn);
+                            }
+                        }
+
+                        // Get columns for each table
+                        List<Column> columnList = new ArrayList<>();
+                        try (ResultSet columns = metaData.getColumns(null, tableSchema, tableName, "%")) {
+                            while (columns.next()) {
+                                String columnName = columns.getString("COLUMN_NAME");
+                                String columnType = columns.getString("TYPE_NAME");
+                                int columnSize = columns.getInt("COLUMN_SIZE");
+                                String nullable = columns.getString("IS_NULLABLE");
+                                columnList.add(new Column(columnName, columnType, columnSize, nullable));
+                            }
+                        }
+                        tableList.add(new Table(tableSchema, tableName, primaryKeyList, columnList));
+                    }
+                }
```
This will get called when on when going to the "Datasource" page (so select Datasources in the side menu). And then again when the dropdown value change to another datasource.

phillip-kruger commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/devui/DatabaseInspector.java:
```diff
@@ -0,0 +1,174 @@
+                .get();
+        AgroalDataSourceSupport agroalSupport = Arc.container().instance(AgroalDataSourceSupport.class)
+                .get();
+        Set<String> excludedNames = support.getInactiveOrHealthCheckExcludedNames();
```
Thanks ! I remove them

phillip-kruger commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/devui/DatabaseInspector.java:
```diff
@@ -0,0 +1,174 @@
+        DataSourceSupport support = Arc.container().instance(DataSourceSupport.class)
+                .get();
+        AgroalDataSourceSupport agroalSupport = Arc.container().instance(AgroalDataSourceSupport.class)
+                .get();
```
Done

phillip-kruger commented:
Ok, there was a few comment from multiple people. I have done the following updates:

@yrodiere

> Note Agroal is not necessarily present if you have a reactive datasource. You might need a similar Dev UI for reactive SQL clients..

Yes, if this is requested we can look at something similar for reactive drivers.

> I know that I would rather work on #39584 than fix an endless stream of bugs related to JDBC quirks -- Hibernate at least hides some of them.

I think we should still have a Hibernate Dev UI Presence (and the log stream :) ) - however for this one the aim is to give a basic DB View. I think we should have both.

What I did:
- Removed the health check excluded names.
- Added a JsonRPC Tests case.
- Added a loading bar when we load the db details.
- Change to page over data - we might still have a problem if the driver loads everything into memory and the DB is massive.

@FroMage

W.r.t HQL queries, this is something we will need to add to the Hibernate extension.
W.r.t byte[] (bytea) like images, I have now added support for those.

![download_binary](https://github.com/user-attachments/assets/6ac61841-472f-4a06-9f08-1f6c4cd47a9e)

@gsmet
- We now page over that result - we might still have a problem if the driver loads everything into memory and the DB is massive.
- bytea are now allowing download (see above)

@Max

> which schema are you listing? the default the connection is for or all of them ?

The schema(s) the user of the connection (as defined in the properties) have access too. We might need to fine tune this for dev service DB. ( I get this from Agroal)

> data security - we already have concerns about exposing config data in devui; but now we will potentially expose a query interface to all data in the raw jdbc connection on the agroal connection? ...should we turn this off by default or at least have option to turn on? (maybe separate devmode only extension?)

There is now a option to turn this off totally (default false) or turn on/off the sql query part (default off). So by default you will not be able to run queries. You need to turn this on in config. Apart from that (after chatting to @yrodiere) I also only allow this feature on local databases (so this feature will not be enabled for datasouces that is not local) 

So not only do we have the LocalHostFilter for dev UI that only allow requests when the host is localhost, but now for this I also made it:

- Local dev only (no remote dev)
- Local DB only - so this will only work if the DB itself is also localhost.

> feature creep - this will open up for massive set of feature requests and can keep us busy forever. We should either agree to deliberately keep this very simple or explore if more suitable reuse of some existing webbased db ui ?

As with anything else in Quarkus, we will make a call on every feature request. The plan is too keep it simple. So real complex things users will still use an external tool (like their IDE)

@cescoffier

- Removed from Menu, as discussed, this is now on the card. I create https://github.com/quarkusio/quarkus/issues/43721 for this
- In another PR we will look (in general in Dev UI) to make sure all JsonRPC services are veto`ed

phillip-kruger commented:
@maxandersen @cescoffier can we make a call on this ?

cescoffier commented:
Where are we on this? Have all the concerns been addressed?

yrodiere commented on extensions/agroal/deployment/src/main/resources/dev-ui/qwc-agroal-datasource.js:
```diff
@@ -0,0 +1,700 @@
+        
+        let jdbcUrl = this._selectedDataSource.jdbcUrl;
+        try {
+            if (jdbcUrl.startsWith("jdbc:h2:mem:") || jdbcUrl.startsWith("jdbc:h2:file:")) {
+                return true;
+            }
+
+            if (jdbcUrl.startsWith("jdbc:derby:memory:")) {
+                return true;
+            }
+
+            if (jdbcUrl.startsWith("jdbc:derby:")) {
+                const derbyUri = jdbcUrl.replace("jdbc:", "");
+                if (derbyUri.startsWith("localhost") || derbyUri.startsWith("127.0.0.1")) {
+                    return true;
+                }
+                if(this._allowedHost && this._allowedHost!="" && derbyUri.startsWith(this._allowedHost)){
+                    return true;
+                }
+            }
+
+            const urlPattern = /^jdbc:[^:]+:\/\/([^:/]+)(:\d+)?/;
+            const match = jdbcUrl.match(urlPattern);
+
+            if (match) {
+                const host = match[1];
+                if(host === "localhost" || host === "127.0.0.1" || host === "::1"){
+                    return true;
+                }
+                if(this._allowedHost && this._allowedHost!="" && host === this._allowedHost){
+                    return true;
+                }
+            }
+
+            return false;
+        } catch (e) {
+            console.error(e);
+            return false;
+        }
+    }
```
So... we're essentially doing client-side security? 

Shouldn't this rather be checked server-side in `DatabaseInspector`?

phillip-kruger replied:
It is also on the server side

yrodiere reviewed (commented):
I'm not sure it's wise to perform "security" checks on the client side, but assuming they get moved to the server side, this seems fine to me.

Now @maxandersen needs to confirm his objections have been addressed.

And I'm not volunteering to maintain this Dev UI; as great as it is, JDBC implies a lot of limitations, so this is only a (great!) first step towards #39584 IMO.

gsmet reviewed (commented):
I won't block it but my gut feeling (that I can't really explain rationally, sorry...) is that we should avoid merging this.

Having seen how hard it has been to develop the clients people use to browse their db, I don't think something simple like that will be solid enough and I foresee it's going to cause some unforeseen issues and we won't be able to make it something more than half baked.

Now again, this is just my gut feeling so feel free to merge it once you are all happy with it. As long as you are certain it's secure.

FroMage commented:
> we won't be able to make it something more than half baked.

It could stay half-baked because we don't have the resources to improve it, but the community might pick it up and turn it into something magical. Chances are that if it's useful, people will improve it.

phillip-kruger commented:
@yrodiere - The client side check is also done on the server side. The client side is just there to prevent a user doing something that we know can not be done (usability rather than security).

@yrodiere - I'll also make a note that you will not maintain this, and make sure you are not being bothered with any bugs / request on this. I'll take care of this as much as possible.

@gsmet - I would not call this half baked. Yes it's not a full blown SQL client, but that's is not what it is. It's a basic table viewer. We can definitely add more features going forward, but I think this is a good start.

yrodiere commented:
> @yrodiere - The client side check is also done on the server side. The client side is just there to prevent a user doing something that we know can not be done (usability rather than security).

Ah, sorry. Indeed I see it now. Thanks for the clarification.

> @gsmet - I would not call this half baked. Yes it's not a full blown SQL client, but that's is not what it is. It's a basic table viewer. We can definitely add more features going forward, but I think this is a good start.

I think the "half baked" comment was not about what you did -- I don't think we could do much better. It's more about the fact that this covers the basics, and that going further, covering the many specific cases of each JDBC driver and how they handle types, is simply not achievable -- or at least, not worth the effort.

> @yrodiere - I'll also make a note that you will not maintain this, and make sure you are not being bothered with any bugs / request on this. I'll take care of this as much as possible.

Ok then. Thanks, and good luck :)

Let's see what Max things of it.

FroMage commented:
> make sure you are not being bothered with any bugs / request on this. I'll take care of this as much as possible

Famous Last Words if I ever heard any 🤣

BTW, just to make that clear, it's a big +1 from me, this PR.

cescoffier reviewed (commented):
I think it would be great to have this one in. The security concerns have been addressed and the usage is restricted. It is a complete database viewer - no, and that's not what we want. 

Let's see if we get feature demands. If it gets out of control, we can decide what to do.

cescoffier commented:
Let's wait until we have the ability to move this code to a "dev only" dependency (@aloubyansky is working on this).

aloubyansky commented:
Here is a draft https://github.com/quarkusio/quarkus/pull/44424

phillip-kruger commented:
@cescoffier @maxandersen This is now updated to use the new dev mode only dependency (https://quarkus.io/version/main/guides/conditional-extension-dependencies#dev-mode-only-extension-dependencies) from @aloubyansky . 

Let's get this merged !

maxandersen commented:
given lots of tests relates to hibernate/persistence/datasource they are probably related to this and should be fixed/investigate before adding.

phillip-kruger commented:
@maxandersen @cescoffier CI error should be fixed now. Just to sum up, we have now used the new conditional dependencies for the Dev UI Runtime code, so that it never end up in prod build. We also made this Dev Mode only and allowed switching this off with a config.

phillip-kruger commented:
Also discussed the use of the dev mode only conditional dependencies with @aloubyansky and all seems good to go.

phillip-kruger committed:
Add a Dev UI Screen for Agroal (DB)

Signed-off-by: Phillip Kruger <phillip.kruger@gmail.com>

phillip-kruger commented:
We have a green CI !!

aloubyansky commented:
As discussed with Phillip, I added a commit on top that configures the resolver for dev mode in `QuarkusDevModeTest`. That's the reason why `quarkus-agroal-dev` wasn't enabled in the `logging-panache` IT tests and had to be added a test scoped dependency.

aloubyansky commented:
Those are flaky test failures. They passed locally for me.

michalvavrik commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourcesJdbcBuildTimeConfig.java:
```diff
@@ -25,6 +28,13 @@ public interface DataSourcesJdbcBuildTimeConfig {
+    /**
+     * Dev UI.
+     */
+    @WithDefaults
```
This annotation is used for maps only according to the docs https://smallrye.io/smallrye-config/Main/config/mappings/#withdefaults. Does it have any function here?

michalvavrik commented on integration-tests/logging-panache/pom.xml:
```diff
@@ -30,6 +30,11 @@
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-agroal-dev</artifactId>
+            <scope>test</scope>
```
Why is this test dependency added here? I can't see any other file or change in this module. Maybe it deserves a comment above the dependency for others to understand why is it present.

aloubyansky replied:
This dependency was removed in a follow up PR https://github.com/quarkusio/quarkus/commit/a84b2852c44480da970a782c2dccda0a02e9a485

michalvavrik replied:
Alright, thank you.

michalvavrik commented on extensions/agroal/runtime/pom.xml:
```diff
@@ -90,6 +90,19 @@
+                        </goals>
+                        <configuration>
+                            <conditionalDevDependencies> 
+                                <artifact>${project.groupId}:${project.artifactId}-dev:${project.version}</artifact> 
```
While it's cool you can do that, the ` <artifactId>quarkus-agroal-dev</artifactId>` is hardcoded anyway and if someone is looking for usage (or simply trying to figure how it works) of the `quarkus-agroal-dev`, they won't find it. If the `artifactId` ever changes, then also will have to change the `quarkus-agroal-dev` so I think this is unnecessary and it would be easier to read if the value was hardcoded.

michalvavrik commented on extensions/agroal/runtime-dev/src/main/java/io/quarkus/agroal/runtime/dev/ui/DatabaseInspector.java:
```diff
@@ -0,0 +1,376 @@
+    private static record Datasource(String name, String jdbcUrl, boolean isDefault) {
+    }
+
+    private static record DataSet(List<String> cols, List<Map<String, String>> data, String error, String message,
```
FYI inner-records are always `static`

michalvavrik commented on extensions/agroal/runtime-dev/src/main/java/io/quarkus/agroal/runtime/dev/ui/DatabaseInspector.java:
```diff
@@ -0,0 +1,376 @@
+                    && !lsql.contains("update ")
+                    && !lsql.contains("delete ")
+                    && !lsql.contains("insert ")
+                    && !lsql.contains("create ")
```
Sorry, this can be stupid question but I didn't try to run this new cool feature yet, what happens if I do `select product_create from products`, this this `contains("create ")` matches?
